wagtail/admin/checks.py

```diff
@@ -150,3 +152,64 @@ def check_panels_in_model(cls, context='model'):
         errors.append(error)
 
     return errors
+
+
+@register()
+def date_format_check(app_configs, **kwargs):
+    errors = []
+
+    date_input_formats = get_format('DATE_INPUT_FORMATS')
+    date_format = getattr(settings, 'WAGTAIL_DATE_FORMAT', None)
+    if date_format is not None and date_format not in date_input_formats:
+        error_hint = (
+            "WAGTAIL_DATE_FORMAT (%r) is not one of the supported formats in "
+            "DATE_INPUT_FORMATS: %r. This may result in validation errors when "
+            "saving forms that include date pickers."
+            % (date_format, date_input_formats)
+        )
+
+        errors.append(
+            Warning(
+                "WAGTAIL_DATE_FORMAT is not a recognised input format",
+                hint=error_hint,
+                id='wagtailadmin.W003',
+            )
+        )
+
+    datetime_input_formats = get_format('DATETIME_INPUT_FORMATS')
```

Comment:
Generally this looks great. However, `get_format` only gets it for the current locale/language - if a user sets up a language preference that isn't the server default in their profile for the Wagtail Admin, I believe it would still be possible for validation to fail without this check catching it. Potentially we could use the `lang=` kwarg to loop over the allowed user profile languages and check `WAGTAIL_*_FORMAT` is consistent with all if `USE_L10N=True`?

Reply:
Aaah, that's annoying - I hadn't made the connection that the *_INPUT_FORMATS lists are liable to vary on a per-user basis. Effectively, this means that the only way to do something not-totally-standard with date formats (e.g. 12-hour times) while USE_L10N is enabled is to supply a custom locale file (as per https://github.com/wagtail/wagtail/issues/4846#issuecomment-433618730) for every locale that a user might conceivably be using.

I think your original tack was the logically correct one then - having WAGTAIL_*_FORMAT configurable as a global setting is fundamentally at odds with USE_L10N. There's a fair argument to be made that anyone who enables Django locale support is disqualifying themselves from being able to say "I want times to display in 12-hour format", unless what they _really_ mean is "I want to provide my users with one or more custom locale options that use 12-hour times" :-) I guess we'll need to take some time to figure out the best compromise between what Django conventions dictate, and what Wagtail users already expect to be able to do.

Reply:
How would you feel about merging the uncontroversial part of this PR (ie minus the `USE_L10N` exception) for now until we make that decision, so we get at least the 12 hour, L10N off, support in 2.10? That way we're not changing Wagtail's behaviour from standard on this issue (though I agree we need a better solution soon, because right now I agree the global setting approach just doesn't work with L10N). Then maybe discuss the L10N issue in the next core team meeting and I can create a separate PR with whatever we decide?

Reply:
Sounds good 👍

Reply:
Great, thanks! Are you happy to tweak and merge or shall I?

Reply:
I'll do that now.